### DISCUSSION TITLE
Missing CFont functions in game_sa

### DISCUSSION THREAD

**Discussion by Disinterpreter** [multitheftauto/mtasa-blue#1895] [Ideas]
Now CFont has only `CFontSA::PrintString` function

But the Internet known more, like:
```
void Font_SetColor(DWORD dwColor);
void Font_SetDropColor(DWORD dwColor);
void Font_SetOutline(int pos);
void Font_Unk12(int unk);
void Font_SetScale(float X, float Y);
void Font_SetJustify(int just);
void Font_SetFontStyle(int style);
void Font_SetProportional(int prop);
void Font_SetRightJustifyWrap(float wrap);
void Font_UseBox(int use, int unk);
void Font_UseBoxColor(DWORD color);
void Font_SetLineWidth(float width);
void Font_SetLineHeight(float height);
void Font_SetShadow(int shadow);
void Font_UnkConv(char *sz, int param1, int param2, int param3, int param4, int param5, int param6, char * buf);
void Font_UnkConv2(char *sz);
```

And we know all addresses and more.

Should we implement it?

And ask me or @saml1er in DM for details ;) because the place where I found it is **** leaked sources (sorry)

P.S It needs for AMX.

**Comment by Disinterpreter** [comment]
After implementing I suppose we have to make functions for Lua. 

What are names of these functions can be? How many functions. It would be good if we discuss about it.

**Comment by Pirulax** [comment]
I personally don't think these are useful in MTA (since we have dxDrawText).
Also, they may confuse some people (That is, they might be wondering: why do we have this, when theres dxDrawText).

**Reply by Disinterpreter** [reply]
> 
> 
> I personally don't think these are useful in MTA (since we have dxDrawText).
> Also, they may confuse some people (That is, they might be wondering: why do we have this, when theres dxDrawText).

Really, I got you. But it's default feature of GTA, and I don't think that we can forget it. It can be useful for developers witch trying to add some retro to them servers.

**Reply by colistro123** [reply]
Not having a basic CFont implementation with callable natives, makes it very difficult to re-create any kind of 'TextDraw'/'PlayerTextDraw' UI for the amx emulator. I tried to do such several times using `dxDraw*` functions, but it never matched the original and it always resulted in frustration to the point where I just gave up on it altogether since I was hitting walls there. Even trying to re-create CFont's class entirely in lua, which was a task I was going to do last year didn't seem very promising to me since there's very specific things related to how the game renders fonts including scaling and the default `MODELS/FONTS.TXD` the game loads in order to render those, driving most of my motivation away. I think adding CFont with a few client-sided natives would be a very simple task, which would save time when trying to re-create these things, rather than 'emulating' UI stuff the game already natively serves.

This is what my 'textdraws' looked like when I tried to emulate them:
![MTA TextDraw attempt](http://whileago.mov.re/nffiles/images/proxy_sa_2020-07-01_18-51-15.jpg)



Something similar to this perhaps:
https://github.com/codenulls/gta-reversed/blob/master/source/game_sa/CFont.cpp

Thanks.

**Reply by Pirulax** [reply]
Well, okay, seems reasonable enough. But isn't AMX separately compiled on client as well?
If no, then it could be added, but please, give me an example API, as I'm lazy to think one out.
Also I have no clue how all this font stuff works, so.

**Comment by Haxardous** [comment]
MTA misses the usage of the O.G. text style, I tried reproducing it but still feel that something is off. It does actually over a set of features which are good to have.

![image](https://user-images.githubusercontent.com/30158626/232549485-0046c636-5182-4696-a3a0-d290f7c57c4c.png)

bringing more base game features into MTA instead of letting people make their own implementation would be great, there are features like the selection menu which are totally forgotten about.

reference:
https://github.com/gta-reversed/gta-reversed-modern/blob/master/source/game_sa/Font.cpp
https://github.com/DK22Pac/plugin-sdk/blob/master/plugin_sa/game_sa/CFont.cpp

**Reply by Haxardous** [reply]
turns out CFontSA & CFont classes have been removed at some point from MTA (probably from the recent removal of unused classes/methods phase)
